File: src/types/questionnaire.ts
export type AnswerScalar = string | number | boolean
export type AnswerValue = AnswerScalar | AnswerScalar[]

export type ModuleOption = {
  label: string
  value: AnswerScalar
  description?: string
  cite?: string
  exclusive?: boolean
}

export type ModuleQuestion = {
  id: string
  text: string
  ref?: string
  description?: string
  type: 'single_choice' | 'multi_choice' | 'multiple_choice' | 'boolean'
  options?: ModuleOption[]
}

export type Module = {
  id: string
  title: string
  description?: string
  questions: ModuleQuestion[]
}

export type StartResponse = {
  session_id: string
  module: Module
}

export type ModuleResponse = {
  module: Module
}

export type QuestionResponse = {
  question: ModuleQuestion
}

export type SubmitRequest = {
  session_id: string
  module_id?: string | null
  answers: Record<string, AnswerValue>
  replace?: boolean
}

export type NextAction = {
  type: 'module' | 'result'
  module_id?: string | null
  message?: string | null
}

export type SubmitResponse = {
  session_id: string
  parameters: Record<string, AnswerValue>
  next: NextAction
  module_complete: boolean
  module?: Module | null
  conclusion?: Record<string, unknown> | null
}

export type ResultResponse = {
  parameters: Record<string, AnswerValue>
  conclusion?: Record<string, unknown> | null
}

File: backend/app/api/routers/questionnaire.py
from fastapi import APIRouter, Depends

from ..schemas import ModuleResponse, QuestionResponse, ResultResponse, StartResponse, SubmitAnswerRequest, SubmitResponse
from ...container import get_service
from ...services.questionnaire import QuestionnaireService

router = APIRouter()


@router.post("/start", response_model=StartResponse)
def start(lang: str = "en", service: QuestionnaireService = Depends(get_service)) -> StartResponse:
    session_id, module = service.start(lang)
    return StartResponse(session_id=session_id, module=module)


@router.get("/module/{module_id}", response_model=ModuleResponse)
def get_module(
    module_id: str, session_id: str, lang: str | None = None, service: QuestionnaireService = Depends(get_service)
) -> ModuleResponse:
    module = service.get_module(session_id, module_id, lang)
    return ModuleResponse(module=module)


@router.post("/submit-answer", response_model=SubmitResponse)
def submit_answer(
    req: SubmitAnswerRequest, lang: str | None = None, service: QuestionnaireService = Depends(get_service)
) -> SubmitResponse:
    payload = service.submit_answer(req.session_id, req.module_id, req.answers, req.replace, lang)
    return SubmitResponse(
        session_id=payload["session_id"],
        parameters=payload["parameters"],
        next=payload["next"],
        module_complete=payload["module_complete"],
        module=payload["module"],
        conclusion=payload["conclusion"],
    )


@router.get("/result", response_model=ResultResponse)
def result(
    session_id: str, lang: str | None = None, service: QuestionnaireService = Depends(get_service)
) -> ResultResponse:
    parameters, conclusion = service.result(session_id, lang)
    return ResultResponse(parameters=parameters, conclusion=conclusion)


@router.get("/question/{question_id}", response_model=QuestionResponse)
def get_question(question_id: str, lang: str | None = None, service: QuestionnaireService = Depends(get_service)) -> QuestionResponse:
    question = service.get_question(question_id, lang)
    return QuestionResponse(question=question)
